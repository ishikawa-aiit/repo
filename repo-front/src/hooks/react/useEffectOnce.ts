import { useEffect, useRef, useState } from 'react'

/**
 * 依存関係がゼロの時、useEffect（）の代わりに使用するカスタムフック
 *
 * @remarks
 * このカスタムフックはReact 18の仕様変更のため、作成しています（React 18では依存関係がゼロの時、2回レンダリングされる）
 * https://github.com/vercel/next.js/issues/35822
 */
export const useEffectOnce = (effect: () => void | (() => void)) => {
  const destroyFunc = useRef<void | (() => void)>()
  const effectCalled = useRef(false)
  const renderAfterCalled = useRef(false)
  const [_val, setVal] = useState<number>(0)

  if (effectCalled.current) {
    renderAfterCalled.current = true
  }

  useEffect(() => {
    if (!effectCalled.current) {
      destroyFunc.current = effect()
      effectCalled.current = true
    }

    setVal((val) => {
      return val + 1
    })

    return () => {
      if (!renderAfterCalled.current) {
        return
      }
      if (destroyFunc.current) {
        destroyFunc.current()
      }
    }
    // マウント時に1回だけ実行するため
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])
}
