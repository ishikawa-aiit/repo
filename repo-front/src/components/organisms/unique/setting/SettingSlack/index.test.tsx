import { screen, render } from '@testing-library/react'

import { SettingSlack } from '.'

describe('props:hasSlackConnection', () => {
  it('trueの場合、設定済みが表示されている', () => {
    const { container } = render(<SettingSlack hasSlackConnection={true} />)

    expect(screen.getByText('設定済み')).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  it('falseの場合、未設定が表示されている', () => {
    const { container } = render(<SettingSlack hasSlackConnection={false} />)

    expect(screen.getByText('未設定')).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })
})
