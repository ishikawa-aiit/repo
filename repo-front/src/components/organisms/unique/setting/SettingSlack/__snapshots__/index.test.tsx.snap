// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`props:hasSlackConnection falseの場合、未設定が表示されている 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-1 {
  width: 6px;
  height: 6px;
  display: inline-block;
  line-height: 1em;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  color: neutral.500;
}

.emotion-2 {
  font-style: italic;
  color: neutral.500;
}

<div>
  <div
    class="emotion-0"
    data-cy="SlackSettingStatus"
  >
    <svg
      aria-hidden="true"
      class="chakra-icon emotion-1"
      fill="none"
      focusable="false"
      stroke="currentColor"
      stroke-width="1"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M9 12l2 2 4-4m6 2a9 9 0 11-18 0 9 9 0 0118 0z"
        stroke-linecap="round"
        stroke-linejoin="round"
      />
    </svg>
    <p
      class="chakra-text emotion-2"
    >
      未設定
    </p>
  </div>
</div>
`;

exports[`props:hasSlackConnection trueの場合、設定済みが表示されている 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-1 {
  width: 6px;
  height: 6px;
  display: inline-block;
  line-height: 1em;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  color: primary.500;
}

.emotion-2 {
  font-style: italic;
  color: primary.500;
}

<div>
  <div
    class="emotion-0"
    data-cy="SlackSettingStatus"
  >
    <svg
      aria-hidden="true"
      class="chakra-icon emotion-1"
      fill="none"
      focusable="false"
      stroke="currentColor"
      stroke-width="1"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M9 12l2 2 4-4m6 2a9 9 0 11-18 0 9 9 0 0118 0z"
        stroke-linecap="round"
        stroke-linejoin="round"
      />
    </svg>
    <p
      class="chakra-text emotion-2"
    >
      設定済み
    </p>
  </div>
</div>
`;
